### DISCUSSION TITLE
Audio Streaming Support

### DISCUSSION THREAD

**Discussion by Kai-Karren** [snakers4/silero-vad#55] [Q&A] [answered]
Hello,

I have to say that I really like Silero VAD and playing around with it.

Aside from loading complete audio files with torchaudio over the provided utility method, I also tried to stream microphone audio using pyaudio into the VADIterator, but I was not successful yet. The problem that I am facing is that I did not manage to convert (binary) audio chunks from pyaudio into PyTorch tensors yet. I tried loading them as byte-like objects which did not work because the audio format could not be identified by the soundfile backend. Storing the audio as a temporary wav file and loading it works, but this is not what I wanted. I am not that familiar with the audio libraries, unfortunately. I also saw that torchaudio has streaming support on the roadmap, but not yet available. I think streaming support either from pyaudio or any other real-time stream would be really cool and would allow more usages of Silero VAD.

Does anyone know how you can convert binary audio chunks into tensors to feed them into the model? Because this is a question about the usage of Silero VAD is posted it here. However, I guess it also depends on the features that torchaudio provides. In case someone has a solution, I would be very thankful.

Regards,
Kai Karren

**Comment by snakers4** [comment, marked as answer]
> he problem that I am facing is that I did not manage to convert (binary) audio chunks from pyaudio into PyTorch tensors yet
> Does anyone know how you can convert binary audio chunks into tensors to feed them into the model? 
> However, I guess it also depends on the features that torchaudio provides.
> In case someone has a solution, I would be very thankful.

`torchaudio` is used here only for demonstration purposes / simplicity. It is by no means necessary of obligatory to use it. You can just work with audio arrays directly in you case.

The answer to your questions is that any audio stream, if uncompressed, mostl likely is just a stream of `int16` values (the common gotcha is that each 2 bytes are one integer, i.e. 16 = 8 * 2).
And any `wav` file is just a 44 byte header + `int16` integers in byte format.

So - you can essentially treat any steam of bytes as sort-of `wav` file already in memory.
Most likely, there is some generator in `pyaudio` you can just directly stick into `VADIterator` with a simple wrapper.

**The only modifications would be (please check the code):**

- Usually you will receive `int16` integer values, whereas our models accept `float32` `[-1, 1]` normalized values, Conversion is just a few lines of code;
- Typically python libraries use numpy arrays, but luckily PyTorch has `.from_numpy()` method;

```
def int2float(sound):
    _sound = np.copy(sound)  # may be not necessary
    abs_max = np.abs(_sound).max()
    _sound = _sound.astype('float32')
    if abs_max > 0:
        _sound *= 1/abs_max
    _sound = _sound.squeeze()  # depends on the use case
    return _sound
```

**Reply by snakers4** [reply]
Also - any PR with working minimalistic downstream usage examples are welcome.
I doubt this can be easily done in colab w/o custom JS, though.

**Reply by Kai-Karren** [reply]
@snakers4 Thank you very much for your detailed answer. Helped a lot :smile:

I think with the information you provided and your conversion method I managed to get it working. I just had to find out how to convert the bytes that pyaudio returns. When you look at it afterward it is not that complicated.
```
audio_int16 = np.frombuffer(audio_chunk_from_pyaudio, np.int16);

audio_float32 = int2float(audio_int16)
batch = streaming_iterator.prepare_batch(torch.from_numpy(audio_float32))
```
I will do more testing and will then probably also share a small example later.

**Reply by Kai-Karren** [reply]
Also in case that I will create a PR with a simple example, where should I put the example? As an own Jupyter Notebook into an examples folder?

**Reply by snakers4** [reply]
> where should I put the example? As an own Jupyter Notebook into an examples folder?

Probably yes
I have no idea yet how to structure the examples
If it will be able to run via colab, then awesome

I am not sure though that you could easily use streaming interfaces from `pyaudio` in vanilla notebooks
I played a bit with `ipywebrtc` - maybe it will help - https://github.com/maartenbreddels/ipywebrtc/blob/master/docs/source/index.rst - but I more or less managed only to record a whole file, not a stream, idk

Also someone contributed this - https://github.com/snakers4/silero-models/blob/master/colab_utils.py

**Reply by Kai-Karren** [reply]
Unfortunately, `pyaudio` works with locally running vanilla notebooks, but not in colab ones because it cannot get microphone access. `ipwebrtc` looks interesting, but also a bit complex at least at first glance. So with it, you managed to get the microphone audio in colab? 

As you wrote earlier for colab I also just found that you have to use JS to get microphone access similar to the colab-utils you shared.

**Reply by Kai-Karren** [reply]
I also found an alternative to pyaudio https://python-sounddevice.readthedocs.io I have not used it myself yet though. Will probably also only work locally I guess.

**Reply by snakers4** [reply]
To be honest I do not remember why I did not use `ipywebrtc` in colab, it was some time ago, maybe it is worth poking just a bit

Only local notebook example is also ok, I assume it will work only on a local machine, not on a headless server over ssh?

**Reply by Kai-Karren** [reply]
Hey, I found the time to finalize the simple streaming example using a local notebook. I have been a bit busy in the last few weeks with university things.

You can check it out here in my fork https://github.com/Kai-Karren/silero-vad/blob/master/examples/pyaudio-streaming/pyaudio-streaming-examples.ipynb

In case you find it good enough to be included as an example in the repo I can create a pull request :slightly_smiling_face: 

About the headless server question, I would assume that too. Maybe it would be possible to forward your audio device(s) over SSH and back. I only know that this works for GUI applications, for the audio part I have no idea.

**Reply by snakers4** [reply]
Hi, yeah, this looks fine!

Maybe just add some meta information for the users in the notebook:

- Library versions you used (maybe just a simple `!pip install lib=version` or similar)
- The platform this is supposed to run on or you have tried to run it on (i.e. x86, Ubuntu 20.04)
- Any additional information (i.e. did you have to perform any system settings, etc)

There is nothing wrong with the way you just plot the outputs, but in your particular case you could have re-used this [function](https://github.com/snakers4/silero-vad/blob/master/utils_vad.py#L546-L570). I believe you can just import the necessary functions (torch hub clones the repo under the hood) by using `hub_dir = torch.hub.get_dir()` to find where it clones it.

Many thanks!

**Reply by Kai-Karren** [reply]
Hi, I have now added the library versions (in the notebook and also included the requirements.txt from my virtualenv), the platform I used (Ubuntu 21.04), if additional set-up is required (no you only have to install the libraries), and a short readme. 

I saw that somebody else already added a similar microphone streaming example in the meantime, however, I think my example is different and could add value, so I opened a PR. https://github.com/snakers4/silero-vad/pull/65 

About your point with using "single_audio_stream", I did not use it or any other of the util methods because I wanted to create a low-level example using just the model basically :slightly_smiling_face:

**Reply by Kai-Karren** [reply]
Also thank you for answering my initial question and discussing it with me :slightly_smiling_face:

**Reply by snakers4** [reply]
I have merged the PR

**Reply by snakers4** [reply]
>  a similar microphone streaming example 

The point there was to provide a boosting example with WebRTC. It is a bit verbose, but it is nice to have several examples fulfilling different purposes


> I wanted to create a low-level example using just the model basically

This is nice, thank you
